tests/components/bsblan/test_climate.py

```diff
@@ -0,0 +1,311 @@
+    mock_hvac_mode = MagicMock()
+    mock_hvac_mode.value = HVACMode.AUTO
+    mock_bsblan.state.return_value.hvac_mode = mock_hvac_mode
+
+    freezer.tick(timedelta(minutes=1))
+    async_fire_time_changed(hass)
+    await hass.async_block_till_done()
+
+    state = hass.states.get(ENTITY_ID)
+    assert state.state == HVACMode.AUTO
+
+    # Test preset_mode
+    mock_hvac_mode.value = PRESET_ECO
+
+    freezer.tick(timedelta(minutes=1))
+    async_fire_time_changed(hass)
+    await hass.async_block_till_done()
+
+    state = hass.states.get(ENTITY_ID)
+    assert state.attributes["preset_mode"] == PRESET_ECO
+
+
+@pytest.mark.parametrize("static_file", ["static.json"])
+@pytest.mark.parametrize(
+    "mode",
+    [HVACMode.HEAT, HVACMode.AUTO, HVACMode.OFF],
+)
+async def test_async_set_hvac_mode(
+    hass: HomeAssistant,
+    mock_bsblan: AsyncMock,
+    mock_config_entry: MockConfigEntry,
+    static_file: str,
+    mode: HVACMode,
+) -> None:
+    """Test setting HVAC mode via service call."""
+    static_data = json.loads(load_fixture(static_file, DOMAIN))
+    with patch.object(
+        mock_bsblan, "static_values", return_value=StaticState.from_dict(static_data)
+    ):
+        await setup_with_selected_platforms(hass, mock_config_entry, [Platform.CLIMATE])
+
+    # Call the service to set HVAC mode
+    await hass.services.async_call(
+        domain=CLIMATE_DOMAIN,
+        service=SERVICE_SET_HVAC_MODE,
+        service_data={ATTR_ENTITY_ID: ENTITY_ID, ATTR_HVAC_MODE: mode},
+        blocking=True,
+    )
+
+    # Assert that the thermostat method was called
+    mock_bsblan.thermostat.assert_called_once_with(hvac_mode=mode)
+    mock_bsblan.thermostat.reset_mock()
+
+
+@pytest.mark.parametrize(
+    ("hvac_mode", "preset_mode", "expected_success"),
+    [
+        (HVACMode.AUTO, PRESET_ECO, True),
+        (HVACMode.AUTO, PRESET_NONE, True),
+        (HVACMode.HEAT, PRESET_ECO, False),
+    ],
+)
+async def test_async_set_preset_mode(
+    hass: HomeAssistant,
+    mock_bsblan: AsyncMock,
+    mock_config_entry: MockConfigEntry,
+    entity_registry: er.EntityRegistry,
+    hvac_mode: HVACMode,
+    preset_mode: str,
+    expected_success: bool,
+    freezer: FrozenDateTimeFactory,
+) -> None:
+    """Test setting preset mode via service call."""
+    await setup_with_selected_platforms(hass, mock_config_entry, [Platform.CLIMATE])
+
+    # patch hvac_mode
+    mock_hvac_mode = MagicMock()
+    mock_hvac_mode.value = hvac_mode
+    mock_bsblan.state.return_value.hvac_mode = mock_hvac_mode
+
+    # Attempt to set the preset mode
+    if not expected_success:
+        with pytest.raises(HomeAssistantError) as exc_info:
+            await hass.services.async_call(
+                CLIMATE_DOMAIN,
+                SERVICE_SET_PRESET_MODE,
+                {ATTR_ENTITY_ID: ENTITY_ID, ATTR_PRESET_MODE: preset_mode},
+                blocking=True,
+            )
+        assert "set_preset_mode_error" in str(exc_info.value)
```

Comment:
```suggestion
        with pytest.raises(HomeAssistantError, matches="Can't set preset mode to eco when HVAC mode is not set to auto"):
            await hass.services.async_call(
                CLIMATE_DOMAIN,
                SERVICE_SET_PRESET_MODE,
                {ATTR_ENTITY_ID: ENTITY_ID, ATTR_PRESET_MODE: preset_mode},
                blocking=True,
            )
```

Reply:
This gives an error is should be match? Also this won't work as there is more in there error message than the message.

Reply:
mine mistake. The error message is somehow deleted on the set preset mode. Now I fixed it and match does work